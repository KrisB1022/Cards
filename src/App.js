import React from "react";
// import logo from './logo.svg';

function App() {
	return (
		<div className="container">
			<header>{/* <img src={logo} className="App-logo" alt="logo" /> */}</header>
		</div>
	);
}

export default App;
