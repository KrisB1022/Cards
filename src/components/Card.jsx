import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { css } from "emotion";

import {
  Button,
  Card as ReactStrapCard,
  CardHeader,
  CardBody,
  CardImg,
  CardText
} from "reactstrap";

const imageStyles = css`
  max-width: 200px;
`;

class Card extends PureComponent {
  cardTextMarkup = ({ label, text }) => {
    return (
      <CardText>
        <span className="font-weight-bolder">{`${label}:`}</span> {text}
      </CardText>
    );
  };

  addToUserCards = () => {
    const { card, updateUserCards } = this.props;
    updateUserCards({ card });
  };

  removeFromUserCards = () => {
    const { card, updateUserCards } = this.props;
    updateUserCards({ card, remove: true });
  };

  render() {
    const { card, isUserCard } = this.props;
    const { imageUrl, name, set, text } = card;

    return (
      <div className="col-md-3 flex-grow-1 mb-4">
        <ReactStrapCard className="pl-0 pr-0 h-100">
          <CardHeader>{name}</CardHeader>

          <CardBody className="d-flex flex-column">
            <figure className="figure mx-auto d-block mb-auto border-bottom pb-2">
              <CardImg
                top
                className={`mx-auto d-block figure-img img-fluid ${imageStyles} flex-grow-1`}
                src={imageUrl}
                alt={`${name} card image`}
              />
              <figcaption className="figure-caption text-right">
                {name}
              </figcaption>
            </figure>

            <div className="pt-2 pb-2">
              {text &&
                this.cardTextMarkup({
                  label: "Card info",
                  text
                })}

              {set &&
                this.cardTextMarkup({
                  label: "Set Name",
                  text: set.name
                })}
            </div>

            {isUserCard ? (
              <Button color="secondary" onClick={this.removeFromUserCards}>
                Remove from my cards
              </Button>
            ) : (
              <Button color="primary" onClick={this.addToUserCards}>
                Add to my cards
              </Button>
            )}
          </CardBody>
        </ReactStrapCard>
      </div>
    );
  }
}

Card.propTypes = {
  card: PropTypes.object.isRequired,
  isUserCard: PropTypes.bool
};

Card.defaultProps = {
  imageUrl: "https://via.placeholder.com/223x310.png"
};

export default Card;
