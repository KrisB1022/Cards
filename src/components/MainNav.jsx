import React, { PureComponent } from "react";
import PropTypes from "prop-types";

import { Badge, Button, Navbar, NavbarBrand } from "reactstrap";
import logo from "../images/magic-banner-730x280.jpg";

class MainNav extends PureComponent {
  state = {
    showUserCards: false
  };

  showUserCards = () => {
    const { userCards, getCards } = this.props;

    getCards({
      clearCards: true,
      query: {
        id: Object.keys(userCards.byId)
      }
    });
    this.toggleUserCards();
  };

  hideUserCards = () => {
    this.props.applyFilters();
    this.toggleUserCards();
  };

  toggleUserCards = () =>
    this.setState(prevState => ({
      showUserCards: !prevState.showUserCards
    }));

  render() {
    const { showUserCards } = this.state;
    const { userCards, totalCount } = this.props;
    const numCards = Object.keys(userCards.byId).length;
    const hasNoCards = numCards === 0;

    return (
      <Navbar color="light" light expand="md" className="mb-3 sticky-top">
        <div className="container">
          <NavbarBrand href="/">
            <img
              className="img-fluid"
              src={logo}
              alt="Magic the Gathering logo"
              width={175}
            />
          </NavbarBrand>

          <p className="m-0">{`${totalCount} Cards`}</p>

          {showUserCards ? (
            <Button color="secondary" onClick={this.hideUserCards}>
              Hide my cards <Badge color="light">{numCards}</Badge>
            </Button>
          ) : (
            <Button
              color="primary"
              onClick={this.showUserCards}
              disabled={hasNoCards}
            >
              Show my cards <Badge color="light">{numCards}</Badge>
            </Button>
          )}
        </div>
      </Navbar>
    );
  }
}

MainNav.propTypes = {
  totalCount: PropTypes.number.isRequired,
  userCards: PropTypes.shape({
    byId: PropTypes.object.isRequired
  }).isRequired
};

export default MainNav;
