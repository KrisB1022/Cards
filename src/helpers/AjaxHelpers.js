export function fetchApi(url, options = {}) {
  return fetch(url, {
    mode: "cors",
    method: "GET",
    ...options
  }).then(res => {
    if (res.ok || res.status === 404) {
      return Promise.resolve(res);
    }

    return Promise.reject(res);
  });
}

export function queryFormatHelper(queryParams = {}) {
  const keys = Object.keys(queryParams);

  if (keys.length === 0) {
    return "";
  }

  return keys.reduce((acc, key, index) => {
    const prefix = index === 0 ? "?" : "&";
    const param = `${key}=${encodeURIComponent(queryParams[key])}`;

    return `${acc}${prefix}${param}`;
  }, "");
}
