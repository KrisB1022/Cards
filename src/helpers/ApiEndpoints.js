export const baseUrl = "https://api.magicthegathering.io";

export const cardsEndpoint = `${baseUrl}/v1/cards`;
export const typesEndpoint = `${baseUrl}/v1/types`;
