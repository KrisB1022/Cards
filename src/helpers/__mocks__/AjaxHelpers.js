export const fetchApi = jest.fn();
export const queryFormatHelper = jest.fn();
