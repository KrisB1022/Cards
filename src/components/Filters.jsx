import React, { Fragment, PureComponent } from "react";
import PropTypes from "prop-types";
import { css } from "emotion";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faBars } from "@fortawesome/free-solid-svg-icons";
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
  Modal,
  ModalFooter
} from "reactstrap";

import { fetchApi } from "../helpers/AjaxHelpers";
import { typesEndpoint } from "../helpers/ApiEndpoints";

const buttonStyles = css`
  bottom: 0;
  right: 0;
`;

class Filters extends PureComponent {
  static orderByOptions = ["artist", "name", "rarity", "set", "type"];
  static pageSizeOptions = [10, 20, 50, 100];

  state = {
    isOpen: false,
    typesOptions: []
  };

  componentDidMount() {
    return fetchApi(typesEndpoint)
      .then(async res => {
        const { types } = await res.json();

        this.setState({
          typesOptions: types
        });
      })
      .catch(err => console.error(err)); // TODO: hook up to page toast
  }

  toggle = () => this.setState(prevState => ({ isOpen: !prevState.isOpen }));

  render() {
    const {
      disabled,
      filters,
      onChange,
      onInput,
      resetFilters,
      totalCount
    } = this.props;
    const { isOpen, typesOptions } = this.state;

    const { artist, name, orderBy, pageSize, types } = filters;

    return (
      <Fragment>
        <Button
          color="secondary"
          className={`position-fixed mb-3 mr-3 ${buttonStyles}`}
          disabled={disabled}
          onClick={this.toggle}
        >
          <FontAwesomeIcon icon={faBars} />
        </Button>

        <Modal
          autoFocus={true}
          isOpen={isOpen}
          toggle={this.toggle}
          centered={true}
        >
          <Form onSubmit={evt => evt.preventDefault()} className="p-3">
            <FormGroup>
              <Label for="name">Card Name</Label>
              <Input value={name} name="name" id="name" onInput={onInput} />
            </FormGroup>

            <FormGroup>
              <Label for="artist">Artist</Label>
              <Input
                value={artist}
                name="artist"
                id="artist"
                onInput={onInput}
              />
            </FormGroup>

            <FormGroup>
              <Label for="types">Types</Label>
              <Input
                disabled={disabled}
                type="select"
                name="types"
                onChange={onChange}
                id="types"
                value={types}
              >
                {typesOptions &&
                  typesOptions.map(typesOption => {
                    return (
                      <option
                        key={typesOption}
                        value={typesOption.toLowerCase()}
                      >
                        {typesOption}
                      </option>
                    );
                  })}
              </Input>
            </FormGroup>

            <FormGroup>
              <Label for="orderBy">Order By</Label>
              <Input
                disabled={disabled}
                type="select"
                name="orderBy"
                onChange={onChange}
                id="orderBy"
                value={orderBy}
                className="text-capitalize"
              >
                <option value="" defaultValue disabled />

                {Filters.orderByOptions.map(orderByOption => {
                  return (
                    <option key={orderByOption} value={orderByOption}>
                      {orderByOption}
                    </option>
                  );
                })}
              </Input>
            </FormGroup>

            <FormGroup>
              <Label for="pageSize">Max Results</Label>
              <Input
                disabled={disabled}
                type="select"
                name="pageSize"
                onChange={onChange}
                id="pageSize"
                value={pageSize}
              >
                {Filters.pageSizeOptions.map(pageSizeOption => {
                  return (
                    <option key={pageSizeOption} value={pageSizeOption}>
                      {pageSizeOption}
                    </option>
                  );
                })}
              </Input>
            </FormGroup>
          </Form>

          <ModalFooter>
            <p className="m-0 mr-auto">{`${totalCount} Total Cards`}</p>
            <Button
              disabled={disabled}
              className="ml-3"
              color="link"
              onClick={resetFilters}
            >
              Reset
            </Button>
            <Button className="ml-3" color="secondary" onClick={this.toggle}>
              Close
            </Button>
          </ModalFooter>
        </Modal>
      </Fragment>
    );
  }
}

Filters.propTypes = {
  onChange: PropTypes.func.isRequired,
  onInput: PropTypes.func.isRequired,
  resetFilters: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  filters: PropTypes.shape({
    artist: PropTypes.string,
    name: PropTypes.string,
    orderBy: PropTypes.string,
    pageSize: PropTypes.number,
    types: PropTypes.string
  }),
  totalCount: PropTypes.number
};

export default Filters;
