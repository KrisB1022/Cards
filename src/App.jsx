import React, { Fragment, PureComponent } from "react";

import { Spinner } from "reactstrap";

import { cardsEndpoint } from "./helpers/ApiEndpoints";
import { fetchApi, queryFormatHelper } from "./helpers/AjaxHelpers";
import MainNav from "./components/MainNav";
import Card from "./components/Card";
import Filters from "./components/Filters";

export const initialFilters = {
  orderBy: "name",
  page: 1,
  pageSize: 20,
  types: "creature"
};

class App extends PureComponent {
  state = {
    cards: [],
    userCards: {
      // order: [], // TODO: Look if ordering if good feature to have
      byId: {}
    },
    isLoadingMore: false,
    isPageLoading: true,
    totalCount: 0,
    filters: {
      ...initialFilters
    }
  };

  componentDidMount() {
    return this.getCards({ clearCards: true }).then(() => {
      window.addEventListener("scroll", this.handleScroll);
    });
  }

  handleScroll = () => {
    const { cards, isLoadingMore, totalCount } = this.state;
    const hasAllCards = totalCount === cards.length;

    if (isLoadingMore || hasAllCards) {
      return;
    }

    const { innerHeight, scrollY } = window;

    if (innerHeight + scrollY >= document.body.offsetHeight - 600) {
      this.getCards(document.body.offsetHeight);
    }
  };

  applyFilters = ({ filter, delay = false } = {}) => {
    this.setState(
      prevState => ({
        filters: {
          ...prevState.filters,
          ...filter,
          page: 1
        }
      }),
      () => {
        if (delay) {
          clearTimeout(this.timer);

          this.timer = setTimeout(() => {
            this.getCards({ clearCards: true });
          }, 500);

          return;
        }

        this.getCards({ clearCards: true });
      }
    );
  };

  resetFilters = () => {
    this.setState(
      {
        filters: { ...initialFilters }
      },
      () => {
        this.getCards({ clearCards: true });
      }
    );
  };

  onChange = ({ target: { id, value } }) => {
    const filter = { [id]: value };
    return this.applyFilters({ filter });
  };

  onInput = ({ target: { id, value } }) => {
    const filter = { [id]: value };
    this.applyFilters({ filter, delay: true });
  };

  getCards = ({ clearCards = false, query = this.state.filters } = {}) => {
    const queryParams = queryFormatHelper(query);

    this.setState({ isLoadingMore: true, isPageLoading: clearCards });

    return fetchApi(`${cardsEndpoint}${queryParams}`)
      .then(async res => {
        if (clearCards) {
          window.scroll({
            top: 0,
            left: 0,
            behavior: "smooth"
          });
        }

        const json = await res.json();

        const { headers } = res;
        const totalCount = headers.get("total-count");
        const hasPages = headers.get("link").includes('rel="next"');

        this.setState(() => {
          const { page, ...filters } = this.state.filters;
          const cards = clearCards
            ? json.cards
            : [...this.state.cards, ...json.cards];

          return {
            cards,
            filters: {
              ...filters,
              page: hasPages ? page + 1 : page
            },
            isLoadingMore: false,
            isPageLoading: false,
            totalCount: JSON.parse(totalCount)
          };
        });
      })
      .catch(err => console.warn("An error occurred", err)); // TODO: Catch error, show toast
  };

  updateUserCards = ({ card, remove }) => {
    const { userCards } = this.state;
    const { id } = card;
    const byId = {
      ...userCards.byId,
      [id]: card
    };

    if (remove) {
      delete byId[id];
    }

    this.setState({
      userCards: {
        ...userCards,
        byId
      }
    });
  };

  render() {
    const {
      cards,
      userCards,
      isLoadingMore,
      isPageLoading,
      filters,
      totalCount
    } = this.state;
    const hasCards = cards && cards.length > 0;

    return (
      <div>
        <MainNav
          applyFilters={this.applyFilters}
          getCards={this.getCards}
          totalCount={totalCount}
          userCards={userCards}
        />

        <div className="container">
          {hasCards ? (
            <Fragment>
              <div className="row">
                {cards.map((card, index) => {
                  const {
                    artist,
                    id,
                    imageUrl,
                    name,
                    originalType,
                    setName
                  } = card;
                  const isUserCard = userCards.byId[id] !== undefined;

                  return (
                    <Card
                      artist={artist}
                      card={card}
                      imageUrl={imageUrl}
                      isUserCard={isUserCard}
                      key={`${id}-${index}`}
                      name={name}
                      originalType={originalType}
                      setName={setName}
                      updateUserCards={this.updateUserCards}
                    />
                  );
                })}
              </div>

              {isLoadingMore && (
                <div className="row">
                  <div className="col d-flex justify-content-center align-items-center mb-5">
                    <Spinner color="secondary" />
                  </div>
                </div>
              )}
            </Fragment>
          ) : (
            <div className="row">
              <div className="col d-flex justify-content-center align-items-center vh-100 mt-n5">
                {isPageLoading ? (
                  <Spinner color="secondary" />
                ) : (
                  <p className="m-0">No Data Found</p>
                )}
              </div>
            </div>
          )}

          <Filters
            disabled={isPageLoading || isLoadingMore}
            filters={filters}
            onChange={this.onChange}
            onInput={this.onInput}
            resetFilters={this.resetFilters}
            totalCount={totalCount}
          />
        </div>
      </div>
    );
  }
}

export default App;
