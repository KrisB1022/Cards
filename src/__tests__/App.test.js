import React from "react";
import { shallow } from "enzyme";
import { Spinner } from "reactstrap";

import { cardsEndpoint } from "../helpers/ApiEndpoints";
import MainNav from "../components/MainNav";
import Card from "../components/Card";
import Filters from "../components/Filters";
import App, { initialFilters } from "../App";

import { fetchApi, queryFormatHelper } from "../helpers/AjaxHelpers";
jest.mock("../helpers/AjaxHelpers");

jest.useFakeTimers();

let wrapper;
const mockCards = [
  {
    artist: "Kit Harington",
    imageUrl: "theNorthWall.jpg",
    name: "Aegon Targaryen",
    originalType: "House Stark",
    setName: "King of the North"
  },
  {
    artist: "Peter Dinklage",
    imageUrl: "handOfQueen.svg",
    name: "Tyrion Lannister",
    originalType: "House Lannister",
    setName: "Hand of the Queen"
  },
  {
    artist: "Iain Glen",
    imageUrl: "targaryenLyfe.jpg",
    name: "Jorah Mormont",
    originalType: "House Targaryen",
    setName: "Ser"
  }
];
const success = {
  headers: {
    get: arg => {
      if (arg === "link") {
        return ["rel=next"];
      }

      if (arg === "total-count") {
        return `${mockCards.length}`;
      }
    }
  },
  json: jest.fn().mockReturnValue({
    cards: mockCards
  })
};

beforeEach(() => {
  fetchApi.mockImplementation(() => {
    return Promise.resolve(success);
  });

  wrapper = shallow(<App />);
});

beforeAll(() => {
  window.scroll = jest.fn();
  window.addEventListener = jest.fn();
});

it("renders without crashing", () => {
  shallow(<App />);
});

it("matches snapshot for initialFilters", () => {
  expect(initialFilters).toMatchSnapshot();
});

it("has 1 <MainNav/>", () => {
  expect(wrapper.find(MainNav).exists()).toBeTruthy();
  expect(wrapper.find(MainNav).length).toBe(1);
});

it("has 1 <Spinner/> if cards = 0 and isPageLoading = true", () => {
  wrapper.setState({
    cards: [],
    isPageLoading: true
  });

  expect(wrapper.find(Spinner).exists()).toBeTruthy();
  expect(wrapper.find(Spinner).length).toBe(1);
});

it('has a "No Data" message if card = 0 and isPageLoading = false', () => {
  wrapper.setState({
    cards: [],
    isPageLoading: false
  });

  expect(wrapper.find("p").text()).toMatch(/^No Data Found$/);
});

it(`has ${mockCards.length} <Card/>`, () => {
  expect(wrapper.find(Card).length).toBe(mockCards.length);
});

it("has 1 <Spinner/> if hasCards and isLoadingMore = true", () => {
  wrapper.setState({
    cards: [...mockCards],
    isLoadingMore: true
  });

  expect(wrapper.find(Spinner).exists()).toBeTruthy();
  expect(wrapper.find(Spinner).length).toBe(1);
});

it("has 1 <Filters/>", () => {
  expect(wrapper.find(Filters).exists()).toBeTruthy();
  expect(wrapper.find(Filters).length).toBe(1);
});

it("has correct starting state", () => {
  const app = new App();

  expect(app.state).toMatchSnapshot();
});

it("calls getCards and add listener on window for componentDidMount", async () => {
  window.addEventListener.mockClear();

  const instance = wrapper.instance();
  const spy = jest
    .spyOn(instance, "getCards")
    .mockImplementation(() => Promise.resolve());

  await instance.componentDidMount();

  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith({ clearCards: true });

  expect(window.addEventListener).toHaveBeenCalled();
  expect(window.addEventListener).toHaveBeenCalledWith(
    "scroll",
    instance.handleScroll
  );
});

describe("handleScroll", () => {
  it("returns if isLoadingMore = true", () => {
    wrapper.setState({
      isLoadingMore: true,
      totalCount: 2000
    });

    const actual = wrapper.instance().handleScroll();
    expect(actual).toBeUndefined();
  });

  it("returns if hasAllCards = true", () => {
    wrapper.setState({
      isLoadingMore: false,
      totalCount: mockCards.length
    });

    const actual = wrapper.instance().handleScroll();
    expect(actual).toBeUndefined();
  });

  it.skip("does nothing if scroll position is not far enough", () => {
    const spy = jest.spyOn(wrapper.instance(), "getCards");

    wrapper.setState({
      isLoadingMore: false,
      totalCount: 2000
    });

    window.innerHeight = 300;
    window.scrollY = 100;
    document.style.height = "1001px";

    wrapper.instance().handleScroll();

    expect(spy).not.toHaveBeenCalled();
  });

  it("calls getCards if scroll position at correct position", () => {
    const spy = jest.spyOn(wrapper.instance(), "getCards");

    wrapper.setState({
      isLoadingMore: false,
      totalCount: 2000
    });
    window.innerHeight = 300;
    window.scrollY = 100;
    document.body.style.height = 1000;

    wrapper.instance().handleScroll();
    expect(spy).toHaveBeenCalledTimes(1);
  });
});

describe("applyFilters", () => {
  it("works without args", () => {
    wrapper.setState({
      filters: {
        page: 10203727219383822,
        something: "here",
        nothing: "there"
      }
    });

    const currState = wrapper.state().filters;

    wrapper.instance().applyFilters();

    expect(wrapper.state().filters).toEqual({
      ...currState,
      page: 1
    });
  });

  it("sets correct state when delay = false", () => {
    const filters = { hereIsAFake: true, page: 5005 };
    wrapper.setState({
      filters
    });

    const filter = {
      filterHere: "here I am"
    };

    wrapper.instance().applyFilters({ filter });

    expect(wrapper.state().filters).toEqual({
      ...filters,
      ...filter,
      page: 1
    });
  });

  it("sets correct state when delay = true", () => {
    const filters = { hereIsAFake: true, page: 5005 };
    wrapper.setState({
      filters
    });

    const filter = {
      filterHere: "here I am"
    };

    wrapper.instance().applyFilters({ filter, delay: true });

    jest.runAllTimers();

    expect(wrapper.state().filters).toEqual({
      ...filters,
      ...filter,
      page: 1
    });
  });
});

it("resets filters and calls getCards on resetFilters", () => {
  wrapper.setState({
    filters: {
      1: true,
      3: false,
      mello: "yellow"
    }
  });

  const instance = wrapper.instance();
  const spy = jest.spyOn(instance, "getCards");

  instance.resetFilters();

  expect(wrapper.state().filters).toEqual(initialFilters);

  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith({
    clearCards: true
  });
});

it("calls applyFilters for onChange", () => {
  const instance = wrapper.instance();
  const applyFiltersSpy = jest.spyOn(instance, "applyFilters");

  const event = {
    target: {
      id: "id",
      value: "value"
    }
  };

  instance.onChange(event);

  expect(applyFiltersSpy).toHaveBeenCalledTimes(1);
  expect(applyFiltersSpy).toHaveBeenCalledWith({
    filter: { id: "value" }
  });
});

it("calls applyFilters for onInput", () => {
  const instance = wrapper.instance();
  const applyFiltersSpy = jest.spyOn(instance, "applyFilters");

  const event = {
    target: {
      id: "id",
      value: "value"
    }
  };

  instance.onInput(event);

  expect(applyFiltersSpy).toHaveBeenCalledTimes(1);
  expect(applyFiltersSpy).toHaveBeenCalledWith({
    filter: { id: "value" },
    delay: true
  });
});

describe("getCards", () => {
  const paramsMock = "hereIsMyParams";

  beforeEach(() => {
    queryFormatHelper.mockReturnValue(paramsMock);
  });

  it("calls queryFormatHelper and fetchApi", () => {
    queryFormatHelper.mockClear();
    fetchApi.mockClear();
    wrapper.instance().getCards();

    const { filters } = wrapper.state();
    expect(queryFormatHelper).toHaveBeenCalledTimes(1);
    expect(queryFormatHelper).toHaveBeenCalledWith(filters);

    expect(fetchApi).toHaveBeenCalledTimes(1);
    expect(fetchApi).toHaveBeenCalledWith(`${cardsEndpoint}${paramsMock}`);
  });

  it("sets correct state if no more pages available", async () => {
    const currentFilter = { filter1: true };
    const page = 2;
    const count = "30";
    wrapper.setState({
      totalCount: 2000,
      filters: {
        ...currentFilter,
        page
      }
    });
    const headers = {
      get: arg => {
        if (arg === "link") {
          return ["rel=prev"];
        }

        if (arg === "total-count") {
          return count;
        }
      }
    };

    fetchApi.mockImplementationOnce(() => {
      return Promise.resolve({
        ...success,
        headers
      });
    });

    await wrapper.instance().getCards();

    const { totalCount, filters } = wrapper.state();
    expect(totalCount).toEqual(parseInt(count));
    expect(filters).toEqual({
      ...currentFilter,
      page
    });
  });

  it("sets correct state if more pages available", async () => {
    const currentFilter = { filter1: true };
    const page = 2;
    const count = 30;
    wrapper.setState({
      totalCount: 2000,
      filters: {
        ...currentFilter,
        page
      }
    });
    const headers = {
      get: arg => {
        if (arg === "link") {
          return ['rel="next"'];
        }

        if (arg === "total-count") {
          return count;
        }
      }
    };

    fetchApi.mockImplementationOnce(() => {
      return Promise.resolve({
        ...success,
        headers
      });
    });

    await wrapper.instance().getCards();

    const { totalCount, filters } = wrapper.state();
    expect(totalCount).toEqual(count);
    expect(filters).toEqual({
      ...currentFilter,
      page: page + 1
    });
  });

  it("calls window.scroll if clearCards = true", async () => {
    window.scroll.mockClear();

    await wrapper.instance().getCards({ clearCards: true });

    expect(window.scroll).toHaveBeenCalledTimes(1);
    expect(window.scroll).toHaveBeenCalledWith({
      top: 0,
      left: 0,
      behavior: "smooth"
    });
  });

  it("does not call window.scroll if clearCards = false", async () => {
    window.scroll.mockClear();

    await wrapper.instance().getCards({ clearCards: false });

    expect(window.scroll).not.toHaveBeenCalled();
  });

  it("sets correct cards state if clearCards = true", async () => {
    wrapper.setState({
      cards: [...mockCards]
    });
    const expected = [
      {
        artist: "Here is an artist",
        imageUrl: "image.png",
        name: "Some Name",
        originalType: "A type",
        setName: "Set Set Set"
      }
    ];

    fetchApi.mockImplementationOnce(() => {
      return Promise.resolve({
        ...success,
        json: jest.fn().mockReturnValue({
          cards: expected
        })
      });
    });
    await wrapper.instance().getCards({ clearCards: true });

    const { cards, isLoadingMore, isPageLoading } = wrapper.state();
    expect(cards).toEqual(expected);
    expect(isLoadingMore).toBeFalsy();
    expect(isPageLoading).toBeFalsy();
  });

  it("sets correct cards state if clearCards = false", async () => {
    wrapper.setState({
      cards: [...mockCards]
    });
    const expected = [
      {
        artist: "Here is an artist",
        imageUrl: "image.png",
        name: "Some Name",
        originalType: "A type",
        setName: "Set Set Set"
      }
    ];

    fetchApi.mockImplementationOnce(() => {
      return Promise.resolve({
        ...success,
        json: jest.fn().mockReturnValue({
          cards: expected
        })
      });
    });
    await wrapper.instance().getCards({ clearCards: false });

    const { cards, isLoadingMore, isPageLoading } = wrapper.state();
    expect(cards).toEqual([...mockCards, ...expected]);
    expect(isLoadingMore).toBeFalsy();
    expect(isPageLoading).toBeFalsy();
  });

  it("catches error", async () => {
    const error = "Rejected!";
    fetchApi.mockImplementationOnce(() => Promise.reject(error));
    const consoleSpy = jest
      .spyOn(console, "warn")
      .mockImplementation(args => args);

    await wrapper.instance().getCards();

    expect(consoleSpy).toHaveBeenCalledTimes(1);
    expect(consoleSpy).toHaveBeenCalledWith("An error occurred", error);

    consoleSpy.mockRestore();
  });
});

describe("updateUserCards", () => {
  const byId = {
    1: { hello: "here" },
    2: { goodbye: "there" },
    3: { needSleep: "please?" }
  };
  const card = {
    id: 2020,
    someData: { hereItIs: true }
  };

  beforeEach(() => {
    wrapper.setState({
      userCards: {
        byId
      }
    });
  });

  it("removes correct userCard if remove = true", () => {
    wrapper.setState({
      userCards: {
        byId: {
          ...byId,
          1010: {
            ...card
          },
          [card.id]: {
            ...card
          }
        }
      }
    });

    wrapper.instance().updateUserCards({
      card,
      remove: true
    });

    expect(wrapper.state().userCards.byId).toEqual({
      ...byId,
      1010: {
        ...card
      }
    });
  });

  it("adds userCard", () => {
    const expected = {
      id: 12345,
      thisIsAMock: "maybe??"
    };

    wrapper.instance().updateUserCards({
      card: expected
    });

    expect(wrapper.state().userCards.byId).toEqual({
      ...byId,
      [expected.id]: {
        ...expected
      }
    });
  });
});
