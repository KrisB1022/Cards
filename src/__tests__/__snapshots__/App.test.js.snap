// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has correct starting state 1`] = `
Object {
  "cards": Array [],
  "filters": Object {
    "orderBy": "name",
    "page": 1,
    "pageSize": 20,
    "types": "creature",
  },
  "isLoadingMore": false,
  "isPageLoading": true,
  "totalCount": 0,
  "userCards": Object {
    "byId": Object {},
  },
}
`;

exports[`matches snapshot for initialFilters 1`] = `
Object {
  "orderBy": "name",
  "page": 1,
  "pageSize": 20,
  "types": "creature",
}
`;
